Add Python support for server SSL cert reloading

cauthu opened the pull request:
now that PR #12644 has merged, i'm recreating PR #13140 which got messy after i rebased (surely i didn't do it right).... just creating this new PR so it's clean

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @cauthu isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -424,6 +424,23 @@ def __init__(self, credentials):
+    type of its instances and its instances exist to be passed to
+    other functions.
+
+    NOTE: do not re-use instances of this class, i.e, always create a
```
This seems like a usability problem. What's the reason for this restriction? What problems arise if it is violated?

cauthu replied:
it's because at https://github.com/grpc/grpc/pull/13188/files#diff-96390deae069919cdf8b9d18d2e92687R46 the `c_cert_config` member of `ServerCertificateConfig` is passed to C-core, which takes over the ownership of the pointer, so i have to null it out. so, we could improve usability--i.e., allow re-use--then we have to make copies of `c_cert_config`, perhaps with a new API in C-core such as `cert_config_dup()` or we have to manually recreate `c_cert_config` in cython after passing it to C-core.
suggestions?

cauthu replied:
sounds good. i'll make cython layer making the copies.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -424,6 +424,23 @@ def __init__(self, credentials):
+    new instance when passing/returning to the library.
+
+    """
+
```
Within the gRPC Python subcodebase we maintain a no-blank-line-between-behavior-doc-string-and-behavior-body style.

cauthu replied:
ah thanks; will fix

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -1252,6 +1269,65 @@ def ssl_server_credentials(private_key_certificate_chain_pairs,
+    """Creates a ServerCertificateConfig for use with an SSL-enabled Server.
+
+    Args:
+      private_key_certificate_chain_pairs: A list of pairs of the form
```
Why a list rather than a sequence, collection, or iterable?

cauthu replied:
just copy-paste from `def ssl_server_credentials`. please let me know what you like to change the doc-string to.

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -1252,6 +1269,65 @@ def ssl_server_credentials(private_key_certificate_chain_pairs,
+    """Creates a ServerCredentials for use with an SSL-enabled Server.
+
+    Args:
+      cert_config_fetcher_cb (callable): a callback that takes no
```
Why one parameter that must behave in two different ways depending on context rather than two different parameters each subject to simpler constraints?

cauthu replied:
are you referring to the fact that during server initialization the callback must return a valid config, vs subsequently it can return a valid config or `None`?
if so, that's because it is required by the C-core: during sever initialization, the C-core has to call the callback to obtain the certificate config, and C-core will error without a valid cert config returned. https://github.com/grpc/grpc/blob/master/include/grpc/grpc_security.h#L396

cauthu replied:
in the two different ways i've implemented the callback (in the `interop` test and in a separate example for my own use case), it doesn't seem that complex. essentially the callback has to determine: "do i have new cert info that i want the server to use, if so, return the new cert info, else return None", and the initialization case is a trivial check.

nevertheless, do you mind specifying the interface you have in mind? is it: python grpc asks for two callbacks: one will be called only once for initialization, and the other will be called for every new client connection? the user can of course choose to, for example, provide the same callable object for both.

thanks!

nathanielmanistaatgoogle replied:
Yes, it is "[gRPC Python] asks for two callbacks: one will be called only once for initialization, and the other will be called for every new client connection".

cauthu replied:
Do you want the two callbacks API no matter what, i.e., regardless of the discussion in https://github.com/grpc/grpc/pull/13188#issuecomment-342924015, or it depends on the outcome of that discussion?
If the former (i.e., no matter what), then I'll start on that right away instead of waiting for the discussion :)
Thanks!

cauthu replied:
Do you think it would help to meet on google hangout for a quick 15 minutes or so to iron out the issue? If you can do that, please let me know a few time slots that works for you and I'll schedule the hangout. Thanks!

cauthu replied:
hmm i guess https://github.com/grpc/grpc/pull/13188#pullrequestreview-75202684 says you `requested changes` i guess you want the API with two callbacks regardless of the outcome of this discussion. so i'll go ahead and make the change.

nathanielmanistaatgoogle replied:
Yes, let's absolutely hang out. How's your tomorrow (Friday)? I'm available 8-11:30 in the morning and 12:30-3:30 in the afternoon.

nathanielmanistaatgoogle replied:
One question I'd like us to address: why pass an `initial_cert_config_fetcher` rather than pass an `initial_cert_config`? Why do this dance of indirection rather than just pass the data?

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -1252,6 +1269,65 @@ def ssl_server_credentials(private_key_certificate_chain_pairs,
+        server will then use for the handshake.
+      require_client_auth: A boolean indicating whether or not to
+        require clients to be authenticated. May only be True if
+        root_certificates is not None.
```
What is root_certificates in this context?

cauthu replied:
i will remove the reference to the `root_certificates`; they are not relevant here, but rather will have to be specified in the cert config that the callback returns.

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/__init__.py:
```diff
@@ -1252,6 +1269,65 @@ def ssl_server_credentials(private_key_certificate_chain_pairs,
+        root_certificates is not None.
+
+    Returns:
+      A ServerCredentials for use with an SSL-enabled
```
Just "A ServerCredentials."? The returned value should be valid for use anywhere that a ServerCredentials is valid for use. No need to call attention at this code location to where those other locations might be.

cauthu replied:
just following the lead of `def ssl_server_credentials`, but i should remove the second sentence then?

cauthu replied:
done: just `A ServerCredentials.` https://github.com/grpc/grpc/pull/13188/commits/67ed381d0a8905537e7b31908336cfe8543838b5

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/README.md:
```diff
@@ -0,0 +1,24 @@
+This tests the ability of gRPC server to rotate its SSL cert for use
```
Rather than a README, just put this prose into the module doc string of the test?

cauthu replied:
sure, moving to `_test.py`

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,411 @@
+
+import atexit
+import os
+import subprocess
```
Why does this test have to use multiple processes? Why can't the new feature be tested in a single process?

cauthu replied:
i originally had these separate programs for "demo"/manual testing before looking into gRPC tests. the feature doesn't have to be tested in separate processes.

to test in a single process, can you please point me to an example that can control two clients? do i use/extend something like `ServerClientMixin` in `cygrpc_test.py`?

cauthu replied:
looks like there's the `_secure_intraop_test.py` that i can build off

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,411 @@
+        import time
+        time.sleep(1)
+
+        print 'things should work...',
```
Please don't include print statements in unit tests?

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,411 @@
+
+# stdout and stderr when the client successfully performs the rpc
+def assert_client_success(stdout, stderr):
+    assert stdout == 'Greeter client received: Hello, you\n', 'unexpected stdout: [{}]'.format(stdout)
```
Please favor the assertSomething methods of unittest.TestCase over using raw assert statements in tests.

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/helloworld_pb2_grpc.py:
```diff
@@ -0,0 +1,47 @@
+import grpc
```
Why include generated code for the test? Can the feature not be tested at a layer of abstraction below the generated code?

cauthu replied:
done... tests no longer using these.

lfaraone commented:
CLA bump

cauthu commented:
high-level description of the new C-core api:

one has two options to create server credentials:
1) by specifying a fixed certificate config, or
2) by specifying a certificate config fetch callback that will be called during server initialization and subsequently on every new client connection, and which should return new certificate config or "unchanged."

specifically: one creates a `grpc_ssl_server_credentials_options *` by using either:

1. `grpc_ssl_server_credentials_create_options_using_config()`
1. `grpc_ssl_server_credentials_create_options_using_config_fetcher()`

then one passes the options object to `grpc_ssl_server_credentials_create_with_options()` and then, as before, passes the credentials object to `add_secure_port()`.

since the fetcher case (2) is key for server cert reload, i prefer to focus on it here in this PR and leave support for (1) for a future PR.

cauthu commented:
@nathanielmanistaatgoogle i've updated the tests to address most of your earlier feedback.

i do have a question: in the tests, i would like to check more specific reason of the connection failure: i.e., whether (a) client rejects server, or (b) server rejects client. i used to check for (a) with
```
        # this assert is fragile, i.e., depends on how the core lib logs the error
        # assert 'SSL_ERROR_SSL: error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED' in stderr
```
when the tests were launching client processes. do you know how i can achieve the same with the new way i'm doing the tests?

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,363 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+'''
```
Small point, but: in the gRPC Python codebase all doc strings are triple-double-quoted, not triple-single-quoted.

cauthu replied:
will fix

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,363 @@
+# limitations under the License.
+
+'''
+This tests the ability of gRPC server to rotate its SSL cert for use
```
The first line of any doc string should be a one-line summary sentence followed by a paragraph break.

cauthu replied:
will update

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,363 @@
+# Copyright 2017 gRPC authors.
```
Why locate this test in tests/interop rather than tests/unit? tests/interop is a bit of a special directory because everything there should have something to do with [the gRPC Interop Test Cases](https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md).

cauthu replied:
will fix... it was just that it wasn't obvious to me where to put these tests. in the other PRs i had solicited feedback on a few possible locations but didn't get any feedback yet, so i just went with interop in the mean time

cauthu replied:
done. moved to `unit`

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,363 @@
+        )
+
+
+class _ServerSSLCertReloadTestBase(unittest.TestCase):
```
Don't put `Base` in the names of classes you intend to use as base classes.

cauthu replied:
will update

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/interop/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,363 @@
+                methods.TestCase.EMPTY_UNARY.test_interoperability(client_stub, None)
+            assert exc.exception.code() == grpc.StatusCode.UNAVAILABLE
+            assert exc.exception.details() == 'Connect Failed', exc.exception.details()
+        return
```
What does this return statement accomplish?

cauthu replied:
will remove

cauthu replied:
done

cauthu commented on src/python/grpcio_tests/tests/unit/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,364 @@
+import grpc
+from src.proto.grpc.testing import test_pb2_grpc
+
+from tests.interop import methods
```
hmm actually regarding your "why put this test in `interop`" feedback, i guess i still have this here after having moved the test to `unit`. i'll try to address the other remaining issues first, and then i'll look into removing the `interop` stuff here if you say i should :) and if i should not use `interop` methods here, please let me know what to replace them with if you know off-hand. thanks!

cauthu replied:
fixed with https://github.com/grpc/grpc/pull/13188/commits/a6cd8906fc5d0de133c42e36b8bf6905256386ff : no longer use `interop.modules`

cauthu commented on src/python/grpcio_tests/tests/unit/server_ssl_cert_reload_test/_test.py:
```diff
@@ -0,0 +1,364 @@
+# limitations under the License.
+
+"""
+This tests server certificate rotation support.
```
the doc-string is slightly stale; will update it

cauthu replied:
done

cauthu commented:
i've addressed most of earlier feedback, except for these two, still waiting for your guidance:
* https://github.com/grpc/grpc/pull/13188#discussion_r147771781
* https://github.com/grpc/grpc/pull/13188#discussion_r148346796

of course my recent commits might require further changes :)

cauthu commented:
> i've addressed most of earlier feedback, except for these two, still waiting for your guidance:
>
> #13188 (comment)
> #13188 (comment)

@nathanielmanistaatgoogle I've just fixed the 2nd one (test in `unit` depending on `interop.modules`), so at this time the remaining major issue is the single callback provided by the user that has to distinguish between being called during server initialization and on subsequent client connections.
Because the core lib calls back into the user's cert config fetcher without providing information about what the "current config" is, the user has two options:
1. Always blindly return the current config according to its best knowledge (i.e., never returns `None`), by querying some other source for the cert material, such as on-disk files, and returning that.
This way, the config fetcher does not need to distinguish between the above two situations. The downside is, even if the cert has not actually changed, returning the cert config will cause the core lib to unnecessarily recreate internal SSL structures, etc. (though likely an insignificant cost compared to reading from disks).
1. Keeps track of some state, e.g., timestamps, cert content, of "the last time the core lib was updated with new config," etc. so that it can tell whether it can simply return `None` (if the core lib's current cert config is already up-to-date) or the latest config.
```
class CertConfigFetcher():
    def __init__(self):
        self.ca_pem = None
        self.cert_pem = None
        # ...
        self.cert_config = None
        self._maybe_update_cert_config()

    def _maybe_update_cert_config(self):
        ca_pem = read_ca_pem()
        cert_pem = read_cert_pem()
        # ...
        if self.ca_pem != ca_pem or self.cert_pem != cert_pem or ...:
            self.cert_config = create_cert_config()

    def __call__(self):
        # self._maybe_update_cert_config() can be called here, or
        # maybe not here but in some periodic fashion

        ret = self.cert_config
        self.cert_config = None # will return None until someone creates a new cert config
        return ret

###
server = grpc.server(...)
creds = grpc.ssl_server_credentials_with_cert_config_fetcher(CertConfigFetcher(), ...)
server.add_secure_port(..., creds)
```
This `CertConfigFetcher` class here does not have to go out of its way to distinguish between the initialization call or subsequent calls on new client connections.

What do you think? If you still want a different API, please let me know what it should be, and I'll be happy to implement it.

Thanks!

cauthu commented:
@nathanielmanistaatgoogle please let me know what you think about https://github.com/grpc/grpc/pull/13188#issuecomment-341492803. Much appreciated :)

nathanielmanistaatgoogle commented:
(Thanks for your patience; tried to get this proper attention today but wasn't able to get to it. Tomorrow looks promising.)

cauthu commented:
Sounds good! Thanks!

cauthu commented:
(switched my "primary email" to see if it fixes the cla/linuxfoundation.)

cauthu commented:
@nathanielmanistaatgoogle it is the latter, i.e., gRPC core calls the callback on _every new client_ connection and it is _asking_ the certificate fetcher to make the judgement.
Therefore, (1) the certificate fetcher has to keep some state, in order to be efficient. (2) If it does not keep state, then it has to always query some other source (e.g., disk), for the keys/certs, and always return a new certificate config to the core, which is not efficient because, always having to query that other source, and/or causing the core to re-initialize SSL structures on every client connection.
What do you think?

cauthu commented:
@nathanielmanistaatgoogle I've made the change requested: use two config fetchers. It's a roughly work-in-progress; I want to push it out to get early feedback. I'll work on adding tests in the mean time. Please let me know. Thanks!

cauthu commented:
discussed with Nathaniel offline and we want Python API to ask for an initial cert config (NOT a fetcher) and a cert config fetcher. this way, Python API will insulate the user from having to provide the initial cert config via a fetcher, which is what C core API is.

cauthu commented:
(A) roughly, the new proposed api is:
`ssl_server_credentials_dynamic_cert_config(initial_cert_config, cert_config_fetcher, require_client_auth=False)` where both of those arguments are required.

(B) i was tempted to make it:
`ssl_server_credentials_cert_config(initial_cert_config, cert_config_fetcher[=None], require_client_auth=False)` where only the initial cert config is required, i.e., without specifying the `cert_config_fetcher`, then the server will just use the initial cert config throughout the server's lifetime (which is the current behavior).

suggestions solicited. if (A), we might want to introduce a second API like ``ssl_server_credentials_fixed_cert_config(cert_config, require_client_auth=False)`; however, this should be a separate PR, so as not to hold up this PR :)

cauthu commented:
chatted with Nathaniel. the API is now OK. list of minor items to fix:
* remove shell script in the tests
* DO the TODO in `credentials.pyx.xi`
* squash into one commit
* run all python tests:
```
run_test.py -l sanity # (this covers copyright, lint, yapf)
run_test.py -l python -r sanity
```
* move the test cert hierarchies into `unit/credentials/`, then move the test up to `unit/`
* reorder `ssl_server_certificate_config()` and `ssl_server_credentials_dynamic_cert_config()` in `__init__.py`

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @cauthu isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

cauthu commented:
* sanity for python 2.7
```
$ ./tools/run_tests/run_tests.py -l python --compiler python2.7 -r sanity
Unexpected error getting flaky tests: <type 'exceptions.ImportError'>
PASSED: /home/cauthu/my-grpc-fork/tools/run_tests/helper_scripts/build_python.sh [time=26.7sec, retries=0:0]

Omitting EPOLLEXCLUSIVE tests


2017-11-10 23:56:45,534 detected port server running version 20
2017-11-10 23:56:45,561 my port server is version 20
SUCCESS: All tests passed
$
```

* sanity for python 3.5 (i don't have 3.4 on my machine)
```
$ ./tools/run_tests/run_tests.py -l python --compiler python3.5 -r sanity
Unexpected error getting flaky tests: <type 'exceptions.ImportError'>
PASSED: /home/cauthu/my-grpc-fork/tools/run_tests/helper_scripts/build_python.sh [time=42.3sec, retries=0:0]

Omitting EPOLLEXCLUSIVE tests


2017-11-10 23:54:57,605 detected port server running version 20
2017-11-10 23:54:57,633 my port server is version 20
SUCCESS: All tests passed
$
```

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @cauthu isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @cauthu isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

cauthu commented:
all python 2.7 tests:
```
$ ./tools/run_tests/run_tests.py -l python --compiler python2.7 
Unexpected error getting flaky tests: <type 'exceptions.ImportError'>
PASSED: /home/cauthu/my-grpc-fork/tools/run_tests/helper_scripts/build_python.sh [time=34.0sec, retries=0:0]

Omitting EPOLLEXCLUSIVE tests


2017-11-11 00:14:38,472 detected port server running version 20
2017-11-11 00:14:38,499 my port server is version 20
SUCCESS: All tests passed
$ 
```

all python 3.5 tests (i don't have python 3.4)
```
$ ./tools/run_tests/run_tests.py -l python --compiler python3.5
Unexpected error getting flaky tests: <type 'exceptions.ImportError'>
PASSED: /home/cauthu/my-grpc-fork/tools/run_tests/helper_scripts/build_python.sh [time=37.3sec, retries=0:0]

Omitting EPOLLEXCLUSIVE tests


2017-11-11 00:24:09,243 detected port server running version 20
2017-11-11 00:24:09,271 my port server is version 20
SUCCESS: All tests passed
$ 
```

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @cauthu isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

cauthu commented:
dummy comment to tickle linux foundation cla bot

cauthu commented:
another dummy comment to tickle linux foundation cla bot

cauthu commented:
sorry, one more dummy comment

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+            return ret
+
+
+class _ServerSSLCertReloadTest(object):
```
Eww, yuck, a mixin!

[I'm no fan of implementation inheritance](https://www.youtube.com/watch?v=3MNVP9-hglc), but implementation inheritance is a better choice here than a mixin. Are you able to follow a pattern like
```py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import abc
import unittest

import six


class AbstractSuperClassTest(six.with_metaclass(abc.ABCMeta, unittest.TestCase)):

  def setUp(self):
    print('AbstractSuperClassTest.setUp!')

  def tearDown(self):
    print('AbstractSuperClassTest.tearDown!')


class AppleTest(AbstractSuperClassTest):

  def test_apple(self):
    print('AppleTest.test_apple!')


class BananaTest(AbstractSuperClassTest):

  def setUp(self):
    super(BananaTest, self).setUp()
    print('Banana.setUp!')

  def tearDown(self):
    print('Banana.tearDown!')
    super(BananaTest, self).tearDown()

  def test_banana(self):
    print('BananaTest.test_banana!')


if __name__ == '__main__':
  unittest.main(verbosity=2)
```
?

cauthu replied:
will try!

cauthu replied:
done, or at least getting closer?

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/credentials/cert_hier_1/certs/ca.cert.pem:
```diff
@@ -0,0 +1,31 @@
+-----BEGIN CERTIFICATE-----
```
Avoid abbreviations, acronyms, and initialisms. Can the directories for these files be just "certificate_hierarchy_1" and so forth?

cauthu replied:
of course! will do

cauthu replied:
done. there's the abbreviations in the code, e.g., `def cert_hier_2_root_ca_cert()`; unclear if you want to change those as well?

nathanielmanistaatgoogle replied:
Yes please.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/credentials/cert_hier_1/certs/ca.cert.pem:
```diff
@@ -0,0 +1,31 @@
+-----BEGIN CERTIFICATE-----
```
Even though they're just test data, how would you feel about adding a `README.md` that describes what these files are and what they mean to one another?

cauthu replied:
yep. adding a `README.md`

cauthu replied:
done. added description to existing `README.md` (renamed from `README`).

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+from tests.testing import _server_application
+from tests.testing.proto import services_pb2_grpc
+
+ca_1_pem = resources.cert_hier_1_root_ca_cert()
```
We do not allow global variables. If these fields are constant, their names should be styled appropriately in ALL_CAPITAL_LETTERS_WITH_UNDERSCORES.

cauthu replied:
i'd suspected this would be an issue :) will move inside the base class perhaps?

cauthu replied:
done: moved into the base class

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/resources.py:
```diff
@@ -34,3 +34,71 @@ def private_key():
                                          _CERTIFICATE_CHAIN_RESOURCE_PATH)
+
+
+# the following are two disjoint certificate hierarchies 1 and 2,
```
Drop this; we generally dislike "stanza" markings in our source code (with the solution being to reshape the code so that no module needs "stanzas".) (Yes, the `grpc/__init__.py` file is an exception to this, although we're not thrilled with the stanza markings there either...)

cauthu replied:
will do

cauthu replied:
done

cauthu commented on src/python/grpcio/grpc/_cython/_cygrpc/server.pyx.pxi:
```diff
@@ -17,6 +17,36 @@ cimport cpython
+  # We are not catching any exception here, because cython will
+  # happily catch and ignore it, and will log for us, and also the
+  # core lib will continue as if the cert has not changed, which is a
+  # reasonable behavior.
```
thanks for the tidbit on PHP; it is funny and definitely not something I would subscribe to :)
the exceptions i'm referring to are mainly from the user's fetcher callback. It would be totally advocating for crashing; please let me know how to do that, i.e., instruct cython to not catch and ignore exceptions here.

Another option is to manually catch exceptions here and return `GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_ERROR`; however, the outcome would not be materially different, since the C-core will again just log the error and continue as if the certificate has not changed.

nathanielmanistaatgoogle replied:
That second option sounds like the right thing to do - if the Core calls us (and it does) and it allows "the application messed up!" as an answer (and it does) then we should tell it that the application messed up in cases in which the application messed up.

That the Core decides to keep on serving is a decision with which I disagree, but... it's out of scope of this change to fuss over it.

cauthu replied:
will do. something like this?
```
else:
  try:
    cert_config_wrapper = user_cb()
  except:
    return GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_ERROR
  if cert_config_wrapper is None:
    return GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_UNCHANGED
  # what if the cert_config_wrapper is not of the right type? i.e.,
  # an instance of grpc.ServerCertificateConfig?
```
what kind of logging should we / can we do here? none of the `pyx.pxi` files does logging, so i don't have an example to follow :)

nathanielmanistaatgoogle replied:
`except Exception:` rather than `except:`.

I think we should just do plain old ordinary Python logging: `logging.exception(<string message>)`.

cauthu replied:
i've tried `logging.error("blah blah")` but i can't see it in stdout/error of the test, even though i can see other error messages like:
```
stderr:
D1114 00:21:02.601111581   23569 security_connector.cc:641]  Using new server certificate config (0x7f08e000d720).
D1114 00:21:02.646434865   23570 security_connector.cc:688]  No change in SSL server credentials.
Traceback (most recent call last):
  File "/home/cauthu/my-grpc-fork/src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py", line 88, in __call__
    'just an error for fun, should not affect the test')
ValueError: just an error for fun, should not affect the test
Exception ValueError: ValueError('just an error for fun, should not affect the test',) in 'grpc._cython.cygrpc._server_cert_config_fetcher_wrapper' ignored
```

cauthu replied:
tl;dr: done.

so it seems like the logging does work when i run the a server outside of the testing environment: this is in `server.pyx.pxi`:
```
except Exception:
    logging.exception('Error fetching certificate config')
    return GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_FAIL
...
elif not isinstance(cert_config_wrapper, ServerCertificateConfig):
    logging.error('Error fetching certificate config: certificate '
                  'config must be of type grpc.ServerCertificateConfig')
```
* make the config fetcher always raise an exception:
```
$ python foo.py server 1>/dev/null
ERROR:root:Error fetching certificate config
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/server.pyx.pxi", line 35, in grpc._cython.cygrpc._server_cert_config_fetcher_wrapper
  File "foo.py", line 57, in __call__
    raise Exception('ALWAYS RAISE')
Exception: ALWAYS RAISE
E1114 02:10:39.476019806   10443 security_connector.cc:694]  Failed fetching new server credentials, continuing to use previously-loaded credentials.
^CTraceback (most recent call last):
  File "foo.py", line 110, in <module>
    time.sleep(100)
KeyboardInterrupt
$
```
* make the config fetcher always return a `1`:
```
$ python foo.py server 1>/dev/null
ERROR:root:Error fetching certificate config: certificate config must be of type grpc.ServerCertificateConfig
E1114 02:11:23.706952551   10474 security_connector.cc:694]  Failed fetching new server credentials, continuing to use previously-loaded credentials.
^CTraceback (most recent call last):
  File "foo.py", line 111, in <module>
    time.sleep(100)
KeyboardInterrupt
$
```

cauthu replied:
ok so it's probably error on my part that the logging messages did not show up in stderr during the test while they did when i ran programs outside the test: it might have been because i needed to remove `py27` (and then having to recreate it with `virtualenv py27`) before running the tests; otherwise the tests wouldn't pick up the change that i made to add the `logging.exception()` call.

now that i've done that--ensure the tests use the updated code--i can see the expected error messages, e.g.:
```
D1114 03:00:13.398917976    5456 security_connector.cc:688]  No change in SSL server credentials.
ERROR:root:Error fetching certificate config: certificate config must be of type grpc.ServerCertificateConfig, not list
E1114 03:00:13.428699928    5456 security_connector.cc:694]  Failed fetching new server credentials, continuing to use previously-loaded credentials.


Unexpected successes: []
Test failure
FAILED: py27.test.unit._server_ssl_cert_config_test.ServerSSLCertReloadTestWithClientAuth [ret=1, pid=5412, time=1.4sec]
```

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+
+* ServerSSLCertReloadTestCertConfigReuse: tests gRPC Python's ability
+  to deal with user's reuse of ServerCertificateConfig instances.
+
```
will do. I didn't insert these intentionally; I think they are part of emacs' `alt-q`

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+            with self.assertRaises(grpc.RpcError) as exc:
+                client_stub.UnUn(_application_common.UNARY_UNARY_REQUEST)
+            self.assertEqual(exc.exception.code(), grpc.StatusCode.UNAVAILABLE)
+            self.assertEqual(exc.exception.details(), 'Connect Failed')
```
it's in the similar vein as my earlier question about checking `stderr` (for specific cause of a failure, i.e., "client rejects server" vs "server rejects client"): lacking a defined api to get at the specific reason for the failure, i resort to checking the `"Connect Failed"` message, i.e., the channel might have been `UNAVAILABLE` due to some other "unexpected" (by the test) reason.

i'm not saying what i'm doing is appropriate; just trying to be clear what i'm trying to do. other suggestions on that? or just remove the `Connect Failed` check and live with it? :)

cauthu replied:
done: removed the assertion on the exception `details() == 'Connect Failed'`

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+class ServerSSLCertConfigFetcherParamsChecks(unittest.TestCase):
+
+    def test_params_checks(self):
+        for initial_config, subsequent_config_fetcher, exc_msg in (
```
sure. (it used to be list of length 3 :)

cauthu replied:
done

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+        if expect_success:
+            client_stub.UnUn(_application_common.UNARY_UNARY_REQUEST)
+        else:
+            with self.assertRaises(grpc.RpcError) as exc:
```
done

nathanielmanistaatgoogle replied:
Not yet named `rpc_error_context`?

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+        for initial_config, subsequent_config_fetcher, exc_msg in (
+            (None, str, 'initial certificate config must be specified'),
+            (1, '1', 'certificate config fetcher must be callable'),):
+            with self.assertRaises(ValueError) as exc:
```
exception context no longer needed since the it was used only to check the error message, and i'm not doing that any more since as you suggest earlier it's not advisable.

cauthu commented on src/python/grpcio_tests/tests/unit/resources.py:
```diff
@@ -11,7 +11,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""Constants and functions for data used in interoperability testing."""
+"""Constants and functions for data used in unit testing."""
```
done

cauthu commented:
not squashing the commits at this time, with the assumption that it's easier for you to see the changes that reflect the feedback.

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -263,38 +266,32 @@ def _do_common_test_recipe(self):
-        self._do_setUp(False)
+    def __init__(self, *args, **kwargs):
+        super(ServerSSLCertReloadTestWithoutClientAuth, self).__init__(*args, **kwargs)
+        self.with_client_auth = False
```
i'm not super stoked about having to toggle the test `with_client_auth` this way, and about the use of `__init__()`, as it appears (per stackoverflow :)) that very rarely does one need to do that for unit test cases.

nathanielmanistaatgoogle replied:
... so why not make `with_client_auth` a parameter of `do_common_test_recipe`?

cauthu replied:
because the test needs `with_client_auth` at `setUp()` time in order to initialize the server credentials, to start the server.

nathanielmanistaatgoogle replied:
Let's not overthink this. This is an abstract super class. How do abstract super classes ask for things from concrete subclasses? They define abstract methods and they call those abstract methods and subclasses implement those abstract methods.

So define a
```py
@abc.abstractmethod
def with_client_auth(self):
  raise NotImplementedError()
```
in the superclass and implement it in the subclasses?

cauthu replied:
i've gone with
```
def require_client_auth(self):
  return True
```
in the base class

nathanielmanistaatgoogle replied:
Please don't? I don't think there's a single occurrence of overriding a concrete method implementation in gRPC Python (outside of `setUp` and `tearDown`?); let's not start now.

cauthu replied:
sounds good! fixed

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -94,18 +68,45 @@ def __call__(self):
+
+    def __init__(self, *args, **kwargs):
+        super(_ServerSSLCertReloadTest, self).__init__(*args, **kwargs)
+        self.client_key_1_pem = resources.cert_hier_1_client_1_key()
```
What about these constants (they are constants, aren't they?) earns them promotion from module-scope to instance-scope? What was wrong with their being at module-scope?

cauthu replied:
sure; i don't expect them to change during the tests, so i can make them module-scope, with ALL_UPPERCASE_NAMES

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -154,107 +154,110 @@ def _do_one_shot_client_rpc(self,
         del client_stub
 
     def _do_common_test_recipe(self):
```
This method name can be as short as `_test` (right?).

cauthu replied:
done

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,357 @@
+                client_stub.UnUn(_application_common.UNARY_UNARY_REQUEST)
+            self.assertEqual(exception_context.exception.code(), grpc.StatusCode.UNAVAILABLE)
+
+    def _create_client_stub(self,
```
Insufficient use of `self` parameter to justify placing this at instance-scope. Devolve it to module-scope by swapping `port` for `self`. Since it will be called from the class, place it earlier-in-the-file-than ("above") the class.

cauthu replied:
will do

cauthu replied:
done

cauthu commented:
now i see "you're authorized to contribute as an employee" on the linux foundation website -> dummy comment to tickle the cla bot

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_cython/_cygrpc/server.pyx.pxi:
```diff
@@ -14,8 +14,43 @@
+                    'config must be of type grpc.ServerCertificateConfig, '
+                    'not %s' % type(cert_config_wrapper).__name__)
+      return GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_FAIL
+    cert_config = cert_config_wrapper._cert_config
```
Tuck this inside an `else:`.

cauthu replied:
done

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,366 @@
+        if self.client_num in (2, 3):
+            raise ValueError(
+                'just an error for fun, should not affect the test')
+        if self.client_num != self.switch_cert_on_client_num:
```
done

cauthu commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,366 @@
+            return None
+        else:
+            # do the switch
+            ret = self.cert_config
```
done

nathanielmanistaatgoogle commented on src/python/grpcio/grpc/_cython/_cygrpc/credentials.pyx.pxi:
```diff
@@ -254,34 +272,81 @@ def call_credentials_metadata_plugin(CredentialsMetadataPlugin plugin):
+  credentials.cert_config_fetcher = cert_config_fetcher
+  cdef grpc_ssl_server_credentials_options* c_options = NULL
+  c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
+    GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY if force_client_auth else GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
```
This large conditional expression looks like it should be spread across three lines? We don't have a linter running on Cython code but we do try to keep it limited to seventy-nine columns wide.

cauthu replied:
what's line-breaking convention used in grpc python? three lines like this?
1.
```
GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY \
if force_client_auth \
else GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
```
or 2.
```
  cdef grpc_ssl_client_certificate_request_type client_certificate_request = \
      GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY
  if not force_client_auth:
    client_certificate_request = GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE
  c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
    client_certificate_request,
```
or 3.
```
  cdef grpc_ssl_client_certificate_request_type client_certificate_request
  if force_client_auth:
    client_certificate_request = CRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY
  else:
    client_certificate_request = GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE
  c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
    client_certificate_request,
```
thanks!

nathanielmanistaatgoogle replied:
Not 1, because [we never use backslash](https://google.github.io/styleguide/pyguide.html?showone=Line_length#Line_length). Not 2, because we avoid assigning a value to a field until after determining that the value assigned to the field is the right value to assign to the field. 3 looks okay, but I was thinking more along the lines of 4:
```py
c_options = grpc_ssl_server_credentials_create_options_using_config_fetcher(
    GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY
    if force_client_auth else
    GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE,
    c_cert_config)
```
. We almost always keep to the rule "if a conditional expression doesn't fit on one line, refactor it into a proper `if/else`", but I think this is a reasonable place to use the embedding of the conditional expression within the function call to avoid backslash continuation or a local field.

3 is okay too.

cauthu replied:
thanks! fixed

nathanielmanistaatgoogle commented:
@grpc-testing: test this please.

nathanielmanistaatgoogle commented:
@grpc-testing: test this please.

cauthu commented:
i think all the python tests now pass. i've also incorporated the style/formatting suggestions from the `yapf_code.sh` from the `-l sanity` test.
```
$ ./tools/run_tests/run_tests.py -l python --compiler python3.5 -r sanity 2>&1 | tail -n2
2017-11-15 08:45:34,861 PASSED: py35.test._sanity._sanity_test.SanityTest [time=1.2sec, retries=0:0]
2017-11-15 08:45:34,862 SUCCESS: All tests passed
$ 
$ ./tools/run_tests/run_tests.py -l python --compiler python3.5 2>&1 | tail -n2
2017-11-15 08:59:03,017 PASSED: py35.test.unit._cython._no_messages_server_completion_queue_per_call_test.Test [time=42.7sec, retries=0:0]
2017-11-15 08:59:03,017 SUCCESS: All tests passed
$ 
$ 
$ ./tools/run_tests/run_tests.py -l python --compiler python2.7 -r sanity 2>&1 | tail -n2
2017-11-15 13:21:17,702 PASSED: py27.test._sanity._sanity_test.SanityTest [time=1.3sec, retries=0:0]
2017-11-15 13:21:17,702 SUCCESS: All tests passed
$ 
$ ./tools/run_tests/run_tests.py -l python --compiler python2.7 2>&1 | tail -n2
2017-11-15 13:29:09,473 PASSED: py27.test.unit._cython._no_messages_server_completion_queue_per_call_test.Test [time=94.1sec, retries=0:0]
2017-11-15 13:29:09,473 SUCCESS: All tests passed
$ 
```

cauthu commented:
@nathanielmanistaatgoogle please let me know how the cert config reuse test is doing in the latest run... if it's still failing due to the missing `grpc.RpcError`, which i can't reproduce on my machine, your assistance in figuring out the root cause would be greatly appreciated!

cauthu commented:
in case it makes a difference, here are how i run the test and what the output looks like---i'm wondering if google's internal tests are showing the `ImportError` and/or the `Omitting EPOLLEXCLUSIVE tests`?

```bash
$ git status
On branch python-server-cert-reload
Your branch is up-to-date with 'origin/python-server-cert-reload'.
nothing to commit, working tree clean
$ 
$ ./tools/run_tests/run_tests.py -l python --compiler python2.7 -r ServerSSLCertReloadTestCertConfigReuse
Unexpected error getting flaky tests: <type 'exceptions.ImportError'>
PASSED: /home/cauthu/my-grpc-fork/tools/run_tests/helper_scripts/build_python.sh [time=46.8sec, retries=0:0]

Omitting EPOLLEXCLUSIVE tests


2017-11-16 17:05:38,044 detected port server running version 20
2017-11-16 17:05:38,073 my port server is version 20
SUCCESS: All tests passed
$ 
```

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,525 @@
+import threading
+import time
+
+from collections import namedtuple
```
Import only packages and modules (almost always modules!); never classes, functions, or other code elements of modules.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,525 @@
+
+    def tearDown(self):
+        if self.server:
+            self.server.stop(0)
```
`None` rather than `0` here.

nathanielmanistaatgoogle commented on src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py:
```diff
@@ -0,0 +1,525 @@
+                self.assertEqual(exception.code(), grpc.StatusCode.UNAVAILABLE)
+            else:
+                self.assertEqual(actual_response, expected_response)
+                raise Exception('NOT REACHED')
```
Debugging cruft to clean up.
